typing/typecore.ml

```diff
@@ -561,6 +572,200 @@ and build_as_type_aux env p =
   | Tpat_any | Tpat_var _ | Tpat_constant _
   | Tpat_array _ | Tpat_lazy _ -> p.pat_type
 
+(* Constraint solving during typing of patterns *)
+
+let solve_Ppat_poly_constraint ~refine env loc sty expected_ty =
+  let cty, ty, force = Typetexp.transl_simple_type_delayed !env sty in
+  unify_pat_types ~refine loc env ty (instance expected_ty);
+  pattern_force := force :: !pattern_force;
+  match ty.desc with
+  | Tpoly (body, tyl) ->
+      begin_def ();
+      init_def generic_level;
+      let _, ty' = instance_poly ~keep_names:true false tyl body in
+      end_def ();
+      (cty, ty, ty')
+  | _ -> assert false
+
+let solve_Ppat_alias env pat =
+  begin_def ();
+  let ty_var = build_as_type env pat in
+  end_def ();
+  generalize ty_var;
+  ty_var
+
+let solve_Ppat_tuple ~refine loc env length expected_ty =
+  let vars = List.init length (fun _ -> newgenvar ()) in
```

Comment:
Here I think that passing the `length` and calling `combine` in the result is less elegant than passing the input list and using `List.map`. (One can combine inside in the callee, or in the caller, I don't know what is best.)

Reply:
We're trying to avoid passing syntax to these functions (expect for type syntax).
Albeit here it would be only used for its length.

Reply:
if the input type is `'a list`, maybe it's not syntax? Just a more structured representation of an integer.

Reply:
Done as suggested, adding a `(type a)` to ensure that we don't use the contents.